### PR TITLE
Implement `range-default-formatter`

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Towards #2919/WG21-P2286R8/WG21-P2585R1.

The following tests are derived from libc++'s test files:
- `P2286R8_text_formatting_range_map`
- `P2286R8_text_formatting_range_sequence`
- `P2286R8_text_formatting_range_set`
- `P2286R8_text_formatting_range_string`

Note that libc++'s `range_formatter` is currently buggy on the `m` option, see LLVM-90196 and LLVM-94562.

The test `P2286R8_text_formatting_tuple_disambiguation` covers formatting `pair`s and `tuple`s that are also ranges.

Unblocks libcxx tests:
- `std/containers/sequences/vector.bool/vector.bool.fmt/format.functions.format.pass.cpp`
- `std/containers/sequences/vector.bool/vector.bool.fmt/format.functions.vformat.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtdef/parse.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtmap/format.functions.format.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtmap/format.functions.vformat.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtmap/parse.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtset/parse.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtstr/format.functions.format.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtstr/format.functions.vformat.pass.cpp`
- `std/utilities/format/format.range/format.range.fmtstr/parse.pass.cpp`
- `std/input.output/iostream.format/print.fun/includes.compile.pass.cpp`

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
Remaining test failures seem related to #4320. Perhaps I also need to fix it.

**Comment by cpplearner** [comment]
There are a lot of code movements. Can you split them into a separate commit?

**Comment by frederick-vs-ja** [comment]
> There are a lot of code movements. To make code review easier, can you split them into a separate commit?

Done.

File: stl/inc/format [hunk 1]
```diff
@@ -4064,240 +4000,405 @@ private:
...
+    _FormatContext::iterator format(_Map_type& _Rx, _FormatContext& _Ctx) const {
+        return _Underlying.format(_Rx, _Ctx);
+    }
+};
 
-        return _Unescaped_str;
+template <_RANGES input_range _Rng, class _CharT>
+struct _Range_default_formatter<range_format::set, _Rng, _CharT> {
+private:
+    using _Set_type = _Fmt_maybe_const<_Rng, _CharT>; // exposition only
```
Comment (cpplearner) [inline stl/inc/format:4321]:
```suggestion
    using _Set_type = _Fmt_maybe_const<_Rng, _CharT>;
```

File: tests/std/tests/P2286R8_text_formatting_range_map/test.cpp [hunk 2]
```diff
@@ -0,0 +1,1100 @@
...
+// Adaptor
+//
+
+class adaptor {
+    using adaptee = map<int, int>;
+
+public:
+    using key_type    = typename adaptee::key_type;
+    using mapped_type = typename adaptee::mapped_type;
+    using iterator    = typename adaptee::iterator;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_range_map/test.cpp:1021]: Elsewhere, the code was edited to take advantage of Down With Typename.

File: tests/std/tests/P2286R8_text_formatting_range_set/test.cpp [hunk 3]
```diff
@@ -0,0 +1,1773 @@
...
+#if 0
+  locale::global(locale(LOCALE_fr_FR_UTF_8));
+  check(SV("{-42,5, 0, 1,25, 42,5}"), SV("{::L}"), input);
+
+  locale::global(locale(LOCALE_en_US_UTF_8));
+  check(SV("{-42.5, 0, 1.25, 42.5}"), SV("{::L}"), input);
+
+  locale::global(locale::classic());
+#endif
+#endif // TEST_HAS_NO_LOCALIZATION
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_range_set/test.cpp:1011]: We should just drop this.

File: tests/std/tests/P2286R8_text_formatting_range_set/test.cpp [hunk 4]
```diff
@@ -0,0 +1,1773 @@
...
+        "The argument index value is too large for the number of arguments supplied", SV("{:^^{}::>5}"), input);
+    check_exception(
+        "The argument index value is too large for the number of arguments supplied", SV("{:^^{}::>{}}"), input, 12);
+}
+
+template <class CharT, class TestFunction, class ExceptionTest>
+void test_pointer(TestFunction check, ExceptionTest check_exception) {
+    // Note nullptr_t can only be equality compared so not used in a set.
+    test_pointer<CharT>(check, check_exception, unordered_set{static_cast<const void*>(0)});
+    test_pointer<CharT>(check, check_exception, unordered_multiset{static_cast<void*>(0)});
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_range_set/test.cpp:1152]: Starting from `0` is too gross for me to accept in our codebase, changing this to `nullptr` (which doesn't affect the test behavior).

File: tests/std/tests/P2286R8_text_formatting_range_string/test.cpp [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+    typename Container::const_iterator end() const {
+        return str_.end();
+    }
+
+private:
+    Container str_;
+};
+
+template <class Container>
+constexpr range_format format_kind<test_range_format_debug_string<Container>> = range_format::debug_string;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_range_string/test.cpp:148]: `std::` was removed from this partial specialization, but retained on 129, I think we should be consistent.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for this next step towards completing C++23! :tada: :heart_eyes_cat: :green_heart:

**Pull request merged**